libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt

```diff
@@ -0,0 +1,302 @@
+
+        /**
+         * Include code in your compound
+         */
+        operator fun GeneratedCode.unaryPlus() {
+            if (this is CompoundGeneratedCode) {
+                mutableCodeList.addAll(codeList)
+            } else {
+                mutableCodeList.add(this)
+            }
+        }
+
+        operator fun String.unaryPlus() {
+            mutableCodeList.add(inlineCode(this))
+        }
+
+        internal fun build(): GeneratedCode {
+            return composite(mutableCodeList)
+        }
+    }
+
+    object Empty : GeneratedCode by EmptyCode()
+
+    companion object {
+        /**
+         * Create a composition of multiple instances of Generated Code
+         */
+        fun composite(init: Builder.() -> Unit) = Builder().apply(init).build()
+
+        /**
+         * Create a composition of multiple instances of Generated Code
+         */
+        fun composite(iterable: Collection<GeneratedCode>): GeneratedCode = if (iterable.count() == 1)
+            iterable.first()
+        else
+            CompoundGeneratedCode(iterable.toList())
+
+        /**
+         * Create a composition of multiple instances of Generated Code
+         */
+        fun composite(vararg code: GeneratedCode) = composite(code.toList())
+    }
+}
+
+private data class CompoundGeneratedCode(val codeList: List<GeneratedCode>) : InternalGeneratedCode() {
+    override fun StringBuilder.write(indent: Int) {
+        writeJoined(codeList, "\n\n", indent)
+    }
+
+    override fun GeneratedCodeVisitor.visit() {
+        for (code in codeList) {
+            visit(code)
+        }
+    }
+}
+
+/**
+ * References a member that can be used as a type
+ */
+interface IdentifiableMember {
+    val name: String
+    fun imports(): Set<String> = emptySet()
+
+    companion object {
+        private data class KClassIdentifiable(val kClass: KClass<*>) : IdentifiableMember {
+            override val name: String = kClass.simpleName ?: kClass.qualifiedName!!
+            override fun imports() = setOfNotNull(kClass.qualifiedName)
+        }
+
+        private data class NamedIdentifiable(override val name: String) : IdentifiableMember
+
+        operator fun invoke(kClass: KClass<*>): IdentifiableMember = KClassIdentifiable(kClass)
+        operator fun invoke(name: String): IdentifiableMember = NamedIdentifiable(name)
+    }
+}
+
+/**
+ * References an interface that can be implemented
+ */
+interface GeneratedInterface : IdentifiableMember
+
+private data class CastedInterfaceMember(private val member: IdentifiableMember) : GeneratedInterface, IdentifiableMember by member
+
+fun IdentifiableMember.asInterface(): GeneratedInterface = CastedInterfaceMember(this)
+
+/**
+ * Generated code that is a type that can be used as an Argument
+ */
+interface IdentifiableCode : GeneratedCode, IdentifiableMember
+
+/**
+ * Builder for a provided type whose JVM Name can be set
+ */
+interface JVMNameableBuilder {
```

Comment:
Why is it JVM-specific?
also, the function seems unused, so the intent is not clear.

Reply:
This was me preparing for something like Data sciences or the shell type provider's extension function heavy generated code. It will often result in clashing names. So renaming the JVM name for things seems appropriate

Reply:
Removed. Will add it again as a modifier just like override in case it becomes necessary again